In [1]:
import ISLP

ISLP.load_data("College")

Private   Apps  Accept  Enroll  Top10perc  Top25perc  F.Undergrad  \
0       Yes   1660    1232     721         23         52         2885   
1       Yes   2186    1924     512         16         29         2683   
2       Yes   1428    1097     336         22         50         1036   
3       Yes    417     349     137         60         89          510   
4       Yes    193     146      55         16         44          249   
..      ...    ...     ...     ...        ...        ...          ...   
772      No   2197    1515     543          4         26         3089   
773     Yes   1959    1805     695         24         47         2849   
774     Yes   2097    1915     695         34         61         2793   
775     Yes  10705    2453    1317         95         99         5217   
776     Yes   2989    1855     691         28         63         2988   

     P.Undergrad  Outstate  Room.Board  Books  Personal  PhD  Terminal  \
0            537      7440        3300    450      2200   70        78   
1           1227     12280        6450    750      1500   29        30   
2             99     11250        3750    400      1165   53        66   
3             63     12960        5450    450       875   92        97   
4            869      7560        4120    800      1500   76        72   
..           ...       ...         ...    ...       ...  ...       ...   
772         2029      6797        3900    500      1200   60        60   
773         1107     11520        4960    600      1250   73        75   
774          166      6900        4200    617       781   67        75   
775           83     19840        6510    630      2115   96        96   
776         1726      4990        3560    500      1250   75        75   

     S.F.Ratio  perc.alumni  Expend  Grad.Rate  
0         18.1           12    7041         60  
1         12.2           16   10527         56  
2         12.9           30    8735         54  
3          7.7           37   19016         59  
4         11.9            2   10922         15  
..         ...          ...     ...        ...  
772       21.0           14    4469         40  
773       13.3           31    9189         83  
774       14.4           20    8323         49  
775        5.8           49   40386         99  
776       18.1           28    4509         99  

[777 rows x 18 columns]

The following **College** data details information of 777 US colleges (see: <a href="https://hastie.su.domains/ISLP/ISLP_website.pdf">Introduction into Statistical Learning</a>). The Data contains the following columns:

| Variable | Description |
| --- | ----------- |
| *Private*| Public/private indicator |
| *Apps*| Number of applications received |
| *Accept*| Number of applicants accepted |
| *Enroll*| Number of new students enrolled |
| *Top10perc*| New students from top 10 % of high school class |
| *Top25perc*| New students from top 25 % of high school class |
| *F.Undergrad*| Number of full-time undergraduates |
| *P.Undergrad*| Number of part-time undergraduates |
| *Outstate*| Out-of-state tuition |
| *Room.Board*| Room and board costs |
| *Books*| Estimated book costs |
| *Personal*| Estimated personal spending |
| *PhD*| Percent of faculty with Ph.D.s |
| *Terminal*| Percent of faculty with terminal degree |
| *S.F.Ratio*| Student/faculty ratio |
| *perc.alumni*| Percent of alumni who donate |
| *Expend*| Instructional expenditure per student |
| *Grad.Rate*| Graduation rate |

Consider the following exercises: 
* For ease of use, having "." in a columnname is not nice. Rename all columns with a "." in their columnname by exchanging it with a "_". (Hint: DataFrame.rename)
* Create a new variable "Acceptance_Rate" that calculates the number of students accepted in % based on the number of applications and add it to the dataframe.
* Sort the Dataframe by the variables "Private" and "Apps", consecutively in an ascending manner.
* Create a new binary qualitative variable, called "Elite", by binning the "Top10perc" variable into two groups based on whether or not the proportion of students coming from the top 10% of their high school classes exceeds 50%. (Hint: pd.cut function)
* Use the describe() method of to produce a numerical summary of the variables in the data set.
* Group the data by the variable "Private" and investigate the mean for the variables "Acceptance_Rate", "Room_Board", "Books", "Phd" and "Graduation_Rate". What can you see?
* Group the data by "Private" AND "Elite" and consider the same variables. Can you see a difference?

In [2]:
import ISLP
colleges = ISLP.load_data("College")

In [7]:
# For ease of use, having "." in a columnname is not nice. Rename all columns with a "." in their columnname by exchanging it with a "_". (Hint: DataFrame.rename)
import pandas as pd


colleges_pd = pd.DataFrame(colleges)

for column in colleges_pd.columns:
    new_col = column.replace('.', '_')
    # https://stackoverflow.com/a/66607872
    colleges_pd.columns = colleges_pd.columns.str.replace(column, new_col, regex=False)

print(colleges_pd)

    Private   Apps  Accept  Enroll  Top10perc  Top25perc  F_Undergrad  \
0       Yes   1660    1232     721         23         52         2885   
1       Yes   2186    1924     512         16         29         2683   
2       Yes   1428    1097     336         22         50         1036   
3       Yes    417     349     137         60         89          510   
4       Yes    193     146      55         16         44          249   
..      ...    ...     ...     ...        ...        ...          ...   
772      No   2197    1515     543          4         26         3089   
773     Yes   1959    1805     695         24         47         2849   
774     Yes   2097    1915     695         34         61         2793   
775     Yes  10705    2453    1317         95         99         5217   
776     Yes   2989    1855     691         28         63         2988   

     P_Undergrad  Outstate  Room_Board  Books  Personal  PhD  Terminal  \
0            537      7440        3300    450    

In [4]:
# Create a new variable "Acceptance_Rate" that calculates the number of students accepted in % based on the number of applications and add it to the dataframe.

colleges_pd["Acceptance_Rate"] = (colleges_pd["Accept"]/colleges_pd["Apps"]*100)
print(colleges_pd)

    Private   Apps  Accept  Enroll  Top10perc  Top25perc  F_Undergrad  \
0       Yes   1660    1232     721         23         52         2885   
1       Yes   2186    1924     512         16         29         2683   
2       Yes   1428    1097     336         22         50         1036   
3       Yes    417     349     137         60         89          510   
4       Yes    193     146      55         16         44          249   
..      ...    ...     ...     ...        ...        ...          ...   
772      No   2197    1515     543          4         26         3089   
773     Yes   1959    1805     695         24         47         2849   
774     Yes   2097    1915     695         34         61         2793   
775     Yes  10705    2453    1317         95         99         5217   
776     Yes   2989    1855     691         28         63         2988   

     P_Undergrad  Outstate  Room_Board  Books  Personal  PhD  Terminal  \
0            537      7440        3300    450    

In [5]:
# Sort the Dataframe by the variables "Private" and "Apps", consecutively in an ascending manner.

colleges_pd = colleges_pd.sort_values(by = ['Private', 'Apps'], ascending=[True, True])
print(colleges_pd)

    Private   Apps  Accept  Enroll  Top10perc  Top25perc  F_Undergrad  \
355      No    233     233     153          5         12          658   
672      No    285     280     208         21         43         1140   
706      No    368     344     212          5         27          863   
165      No    434     412     319         10         30         1376   
630      No    441     369     172         17         45          633   
..      ...    ...     ...     ...        ...        ...          ...   
407     Yes  13594    7244    2505         70         86        12408   
174     Yes  13789    3893    1583         90         98         6188   
250     Yes  13865    2165    1606         90        100         6862   
614     Yes  14446   10516    3252         22         57        14130   
59      Yes  20192   13007    3810         45         80        14971   

     P_Undergrad  Outstate  Room_Board  Books  Personal  PhD  Terminal  \
355           58      4486        2516    600    

In [26]:
# Create a new binary qualitative variable, called "Elite", by binning the "Top10perc" variable into two groups based on whether or not the proportion of students coming from
# the top 10% of their high school classes exceeds 50%. (Hint: pd.cut function)

# https://stackoverflow.com/a/54343112

bins = [0, 51, 100]
labels = ["No", "Yes"]
colleges_pd['Elite'] = pd.cut(colleges_pd['Top10perc'], bins=bins, labels=labels)
colleges_pd.tail()

Private   Apps  Accept  Enroll  Top10perc  Top25perc  F_Undergrad  \
772      No   2197    1515     543          4         26         3089   
773     Yes   1959    1805     695         24         47         2849   
774     Yes   2097    1915     695         34         61         2793   
775     Yes  10705    2453    1317         95         99         5217   
776     Yes   2989    1855     691         28         63         2988   

     P_Undergrad  Outstate  Room_Board  Books  Personal  PhD  Terminal  \
772         2029      6797        3900    500      1200   60        60   
773         1107     11520        4960    600      1250   73        75   
774          166      6900        4200    617       781   67        75   
775           83     19840        6510    630      2115   96        96   
776         1726      4990        3560    500      1250   75        75   

     S_F_Ratio  perc_alumni  Expend  Grad_Rate  Acceptance_Rate Elite  
772       21.0           14    4469         40        68.957670    No  
773       13.3           31    9189         83        92.138846    No  
774       14.4           20    8323         49        91.320935    No  
775        5.8           49   40386         99        22.914526   Yes  
776       18.1           28    4509         99        62.060890    No

In [29]:
# Use the describe() method of to produce a numerical summary of the variables in the data set.

colleges_pd.describe()

Apps        Accept       Enroll   Top10perc   Top25perc  \
count    777.000000    777.000000   777.000000  777.000000  777.000000   
mean    3001.638353   2018.804376   779.972973   27.558559   55.796654   
std     3870.201484   2451.113971   929.176190   17.640364   19.804778   
min       81.000000     72.000000    35.000000    1.000000    9.000000   
25%      776.000000    604.000000   242.000000   15.000000   41.000000   
50%     1558.000000   1110.000000   434.000000   23.000000   54.000000   
75%     3624.000000   2424.000000   902.000000   35.000000   69.000000   
max    48094.000000  26330.000000  6392.000000   96.000000  100.000000   

        F_Undergrad   P_Undergrad      Outstate   Room_Board        Books  \
count    777.000000    777.000000    777.000000   777.000000   777.000000   
mean    3699.907336    855.298584  10440.669241  4357.526384   549.380952   
std     4850.420531   1522.431887   4023.016484  1096.696416   165.105360   
min      139.000000      1.000000   2340.000000  1780.000000    96.000000   
25%      992.000000     95.000000   7320.000000  3597.000000   470.000000   
50%     1707.000000    353.000000   9990.000000  4200.000000   500.000000   
75%     4005.000000    967.000000  12925.000000  5050.000000   600.000000   
max    31643.000000  21836.000000  21700.000000  8124.000000  2340.000000   

          Personal         PhD    Terminal   S_F_Ratio  perc_alumni  \
count   777.000000  777.000000  777.000000  777.000000   777.000000   
mean   1340.642214   72.660232   79.702703   14.089704    22.743887   
std     677.071454   16.328155   14.722359    3.958349    12.391801   
min     250.000000    8.000000   24.000000    2.500000     0.000000   
25%     850.000000   62.000000   71.000000   11.500000    13.000000   
50%    1200.000000   75.000000   82.000000   13.600000    21.000000   
75%    1700.000000   85.000000   92.000000   16.500000    31.000000   
max    6800.000000  103.000000  100.000000   39.800000    64.000000   

             Expend  Grad_Rate  Acceptance_Rate  
count    777.000000  777.00000       777.000000  
mean    9660.171171   65.46332        74.692771  
std     5221.768440   17.17771        14.710392  
min     3186.000000   10.00000        15.448631  
25%     6751.000000   53.00000        67.564745  
50%     8377.000000   65.00000        77.875000  
75%    10830.000000   78.00000        84.852217  
max    56233.000000  118.00000       100.000000

In [24]:
# Group the data by the variable "Private" and investigate the mean for the variables "Acceptance_Rate", "Room_Board", "Books", "Phd" and "Graduation_Rate". What can you see?

ar = colleges_pd.groupby(['Private'])["Acceptance_Rate"].mean()
rb = colleges_pd.groupby(['Private'])["Room_Board"].mean()
b = colleges_pd.groupby(['Private'])["Books"].mean()
phd = colleges_pd.groupby(['Private'])["PhD"].mean()
gr = colleges_pd.groupby(['Private'])["Grad_Rate"].mean()
print(ar, rb, b, phd, gr)

Private
No     72.653048
Yes    75.458118
Name: Acceptance_Rate, dtype: float64 Private
No     3748.240566
Yes    4586.143363
Name: Room_Board, dtype: float64 Private
No     554.377358
Yes    547.506195
Name: Books, dtype: float64 Private
No     76.834906
Yes    71.093805
Name: PhD, dtype: float64 Private
No     56.042453
Yes    68.998230
Name: Grad_Rate, dtype: float64


In [28]:
# Group the data by "Private" AND "Elite" and consider the same variables. Can you see a difference?

ar = colleges_pd.groupby(['Private', 'Elite'])["Acceptance_Rate"].mean()
rb = colleges_pd.groupby(['Private', 'Elite'])["Room_Board"].mean()
b = colleges_pd.groupby(['Private', 'Elite'])["Books"].mean()
phd = colleges_pd.groupby(['Private', 'Elite'])["PhD"].mean()
gr = colleges_pd.groupby(['Private', 'Elite'])["Grad_Rate"].mean()
print(ar, rb, b, phd, gr)

Private  Elite
No       No       73.858531
         Yes      54.199882
Yes      No       78.092048
         Yes      53.289211
Name: Acceptance_Rate, dtype: float64 Private  Elite
No       No       3696.376884
         Yes      4542.153846
Yes      No       4471.958416
         Yes      5547.200000
Name: Room_Board, dtype: float64 Private  Elite
No       No       550.005025
         Yes      621.307692
Yes      No       542.237624
         Yes      591.850000
Name: Books, dtype: float64 Private  Elite
No       No       76.175879
         Yes      86.923077
Yes      No       68.780198
         Yes      90.566667
Name: PhD, dtype: float64 Private  Elite
No       No       54.643216
         Yes      77.461538
Yes      No       67.025743
         Yes      85.600000
Name: Grad_Rate, dtype: float64
